Support for emulation mode in IBRDTN

fguidec opened the pull request:
In emulation mode, several instances of IBRDTN can run concurrently on the same host, and all traffic is sent (as unicast traffic) to a software hub, which forwards messages to their actual destination(s). This software hub can itself be controlled by a mobility simulator, in which case each message is only forwarded by the hub if the sender is considered as a neighbor of the destination at the time the message is received by the hub.

LEPTON (http://www-casa.irisa.fr/lepton) is a lightweight emulation platform for opportunistic networking that works along that line, and that is meant to be interoperable with several existing opportunistic networking systems (including IBRDTN).

In order for IBRDTN nodes to support emulation mode, the original IBRDTN code must be slightly modified. The modifications required mostly pertain to UDP-based beaconing and to the TCP convergence layer.  Additionally,  a few optional command-line parameters are defined.

None of these modifications compromise the "standard"  behavior of IBRDTN nodes, which can still interact using multicast beacons and direct (i.e., node-to-node) TCP sessions.

morgenroth commented:
Thank you for re-organizing the changes. Now they are much more clearer to me.
Unfortunately, I'm not happy with them.

Most of the work was spent on choosing a random port and get that information. Why not configuring each instance with a unique port? On an simulation system you should be able to determine a port-range which is under your control. The emulator could choose a free port for each instance and simply create a configuration file for each instance. That way, patch efb1df7 to b426cc8 would be not required. The patch c739493 is still required of course.

The last patch bec2b39 is IMHO also not required for a simulation. The start/stop behavior could be easily controlled externally. Let the daemon start by your simulator at what-ever time you like and stop it using kill procedures. What would be wrong with that?

I'm not against useful changes, but these functionality seems to be better in the simulator instead of the daemons to control. Consider, in your approach these functionality must be implemented within each software stack you want to control. If it is part of the simulator you don't have to struggle with maintainers like me. ;-)

fguidec commented:
On 29/09/2017 20:58, Johannes Morgenroth wrote:
> Thank you for re-organizing the changes. Now they are much more clearer to me.
> Unfortunately, I'm not happy with them.

Thank you for these comments on our pull request. I understand that some 
of the proposed changes could make you unhappy, or at least uneasy, but 
I hope I can explain why we saw no alternatives (and hopefully make you 
happy again!).

Before addressing your comments, I'd like to highlight the main points 
of the "philosophy" behind the LEPTON 
<http://casa-irisa.univ-ubs.fr/leptonemulation> platform.

With this platform we would like to make it possible for any user to run 
emulation experiments involving fully operational OppNet systems such as 
IBDRTN, Active-DTN, DoDWAN, etc. This should be possible for a 
"standard" user on a "standard" workstation (typically, a student in a 
CS lab), so it should not require superuser privileges or a special 
runtime environment. This is, by the way, why our platform does not rely 
on a virtualization technology.

With LEPTON it should be possible to run experimentation scenarios 
involving hundreds or thousands of nodes, and most notably scenarios 
based on traces found in CRAWDAD <https://crawdad.org> datasets. In some 
of these datasets the population of nodes is stable, but others (e.g., 
roma/taxi, kth/walkers, epfl/mobility...) describe mobile networks with 
churn: the population of nodes changes continuously, as nodes enter or 
leave the network over time.

So far LEPTON has been used on a single workstation to run scenarios 
involving a couple hundred nodes. Larger scenarios can be run on a 
cluster of workstations.

LEPTON can also support scenarios in which the population of nodes is 
not known in advance. This occurs for example when the population is 
determined by the application itself (e.g., bio-inspired distributed 
application), or by the behavior of real users (e.g., in conference 
demos). In such cases the emulation platform is not responsible for 
creating OppNet nodes. Instead it simply discovers new nodes as and when 
they begin sending beacons (please have a look at LEPTON's howtos 
<http://casa-irisa.univ-ubs.fr/lepton/doc/howtos/lepton_with_dodwandroid.html> 
if you're interested by this mode of operation).


> Most of the work was spent on choosing a random port and get that information.

Yes, indeed.

> Why not configuring each instance with a unique port? On an simulation system
> you should be able to determine a port-range which is under your control. The
> emulator could choose a free port for each instance and simply create a
> configuration file for each instance.

The only way to "reserve" a port in a Unix/Linux system is to bind a 
socket to that port. Reserving a whole port range therefore requires 
creating and binding as many sockets. This approach presents several 
disadvantages.

1) This is a resource-greedy approach, since creating a socket implies 
creating I/O buffers, etc. Besides it is hardly appropriate for 
scenarios involving networks with churn, for in such scenarios it is 
sometimes difficult to determine how many nodes will be running 
simultaneously at any time during the experiment, and thus how many port 
numbers should be "sufficient" to serve all these nodes.

2) The portreserve/portrelease commands available in most Linux 
distributions actually allow to reserve specific ports, but they are 
meant to be used for a small number of privileged ports, and their 
execution requires superuser privileges. This is therefore not an option 
for LEPTON.
Instead LEPTON could actually try to mimick the portreserve/portrelease 
behavior by reserving a whole range of non-privileged ports before 
starting an experiment, and releasing these ports on demand whenever an 
OppNet node is created.
Yet when running scenarios involving networks with churn, LEPTON would 
also have to keep track of the release of allocated ports (with or 
without any friendly notification from the corresponding OppNet node), 
so as to be able to reassign these ports to other nodes. In case of a 
node terminating unexpectedly (which actually happens quite often when 
new code is being tested), the emulator would have to account for the 
actual states of the pending sockets (i.e., states FIN_WAIT1, FIN_WAIT2, 
CLOSE_WAIT...) in order not to assign a port number that is actually not 
available yet.
All this is feasible, of course, but this is exactly what the operating 
system already does to manage TCP and UDP ports. When a program requests 
the creation of a socket with a randomly assigned port number, it 
actually asks to benefit from this work performed by the OS.

Supporting randomly assigned port numbers in a DTN system should not be 
a problem, since the IPND protocol provides for service (i.e., address 
and port) discovery between neighbour nodes.

What makes it slightly difficult in IBRDTN's code is the fact that 
several sockets are actually created for each interface used by the UDP 
or TCP convergence layers, and by the IPND agent. Indeed, one socket is 
created for each non-local address of the interface the IBRDTN node must 
listen to. If this interface has for example an IPv4 address and two 
IPv6 addresses, then three sockets are created, each socket being bound 
to one of these adresses, but to the same port number of course. When 
this port number must be assigned by the OS (random assignment), the 
problem is to make sure that all sockets eventually listen to the same port.
Quite honestly I wonder why you don't simply create one socket 
(listening to the desired port), without binding it to any interface or 
address. Admittedly an IBRDTN may then receive messages (i.e., beacons 
or bundles) from the wrong interface, but would that be such a problem? 
In contrast the code would be way, way simpler. Besides, the "trick" 
used in patches 8ae0791 
<https://github.com/ibrdtn/ibrdtn/pull/220/commits/8ae0791b782fd801c6b27b64fde476afefbd08c3> 
and b426cc8 
<https://github.com/ibrdtn/ibrdtn/pull/220/commits/b426cc87b1b9a280fe7cb6c8b0552701653de9df> 
would not be required. This trick consists in creating a temporary dummy 
socket first, and using its port number to create all other sockets. I 
must admit I'm not very satisfied with this approach, but after 
considering several other options I realized it was the approach that 
required the minimum amount of changes in the current code.
(For a while I thought it was actually possible to use the TCP or UDP 
convergence layers without binding the sockets to a specific interface 
address, but I never managed to make IBRDTN run that way.)


> The last patch bec2b39
> <https://github.com/ibrdtn/ibrdtn/commit/bec2b39a6d3eb029cc8c850af639b9403176d75f>
> is IMHO also not required for a simulation. The start/stop behavior could be
> easily controlled externally. Let the daemon start by your simulator at
> what-ever time you like and stop it using kill procedures. What would be wrong
> with that?

LEPTON can actually start each OppNet node at a desired time, and kill 
it afterwards. This is one of the modes of operations it supports, and 
this is of course simpler as it does not require any "cooperation" from 
the OppNet nodes.
Yet it is not the most accurate, as far as repeatability is concerned. A 
far better control on time events is obtained by telling each node when 
it should actually start and stop running.

Being able to repeat experiments is of course of prime importance in 
simulation and emulation systems. Although an emulation platform such as 
LEPTON is meant to run in real time, users must be able to play the same 
experimentation scenario as many times as needed, while changing some 
execution parameters (e.g., switching from epidemic to PROPHET routing 
in IBRDTN in order to compare how they perform).

When running scenarios that involve a stable population of nodes, LEPTON 
must create all nodes at the very same time. Yet starting each node 
takes time, as it requires loading code and libraries, opening I/O 
sockets and files, etc.
With scenarios involving a couple hundred nodes, we have observed that 
creating all nodes can take several seconds on a powerful workstation. 
Besides these nodes actually start running in a disordered way, although 
the scenario specifies that they should be synchronized. And of course, 
the latency observed on one node during an experimentation run will be 
completely different during the next run.
By telling each node when it should actually start running, we somehow 
recreate that synchronization. Basically LEPTON creates OppNet nodes in 
advance (typically, 10 seconds before, but this can be adjusted in the 
parameters), but it asks each node to wait until a set time to actually 
start running. We've been able to observe that with this approach 
hundreds of nodes can actually start in a very short time window 
(typically, a few tens of ms).

The same observations can be made for node termination, of course. 
Telling each node when it should terminate is more accurate than sending 
this node a kill signal, especially if hundreds of nodes are expected to 
terminate at the same time.

As explained above, LEPTON can drive OppNet nodes without passing them 
start/stop parameters, but the ability to follow a well-defined scenario 
is then diminished, and it is then a lot more difficult to compare the 
log files produced during successive experiment runs.

> in your approach
> these functionality must be implemented within each software stack you want to
> control.

True. But there are not that many candidate stacks anyway. To the best 
of our knowledge there are not more than half a dozen fully-operational, 
openly distributed OppNet systems to date. So adapting these systems for 
LEPTON is not a tremendous task. Besides we'd like to encourage 
developers to do the work, rather than do it ourselves. Instructions on 
how to interface a system with LEPTON are already available 
<http://casa-irisa.univ-ubs.fr/lepton/doc/howtos/emulated_nodes.html>. 
We did the job for IBRDTN because we thought it would be a good 
illustration of how to do it. People at UAB (Barcelona) are currently 
working on the adaptation of their own system aDTN 
<https://github.com/SeNDA-UAB/aDTN-platform>.

Conversely, adapting LEPTON to each and every OppNet system would lead 
us to adopting the lowest baseline (i.e., what is not possible for one 
OppNet system should not be considered as possible for other OppNet 
systems).

> If it is part of the simulator you don't have to struggle with
> maintainers like me. ;-)

Also true.  But I'm not totally desperate yet! Of course we could keep 
distributing our own modified version of IBRDTN 
<http://casa-irisa.univ-ubs.fr/lepton/doc/howtos/lepton_with_ibrdtn.html>, 
but in the end we would frankly prefer our Web page to point directly to 
the official version of IBRDTN.

So I hope we'll finally come to an arrangement.  ;-)

Best regards.

morgenroth commented:
> The only way to "reserve" a port in a Unix/Linux system is to bind a socket to that port.

As you may not know port ranges to use are easy to identify in Linux. Usually, the ports above 32768 are used for dynamic allocations. The port below are not used for other clients and are usually free. You can check that using proc (even without super-user).

```
$ cat /proc/sys/net/ipv4/ip_local_port_range
32768   60999
```

> Yet it is not the most accurate, as far as repeatability is concerned. A far better control on time events is obtained by telling each node when it should actually start and stop running.

Getting determinism is an ambitious goal with real network stack and emulated movement. Since the OppNets (at least IBR-DTN) is not really deterministic due to multithreading, the determinism of your results will be limited. However, getting all nodes in sync at start is an good point. But how do you find out the time needed to start your (thousands) nodes? 10 seconds sounds like a empirical value. I think there is a better way to do it. In [Hydra](https://github.com/morgenroth/hydrasim), the nodes are started without being connected to each other. As soon as all nodes are ready, the discovery channel is opened. In you case, you could easily control that channel using the discovery relay. The same way, you could disconnect the nodes from the network. First detach and then send a kill signal.

In reality, DTN nodes are usually started long before the first contact and stay on (or in sleep) until the next contact. Starting and stopping the software stacks discards important state within the nodes, which is IMHO unrealistic.

> Quite honestly I wonder why you don't simply create one socket (listening to the desired port), without binding it to any interface or address. Admittedly an IBRDTN may then receive messages (i.e., beacons or bundles) from the wrong interface, but would that be such a problem?

This allows IBR-DTN to bind explicitly to specific interfaces with different convergence-layers. If you want to bind to any interface (e.g. *:4556), you can simply set "any" as interface.

fguidec commented:
On 03/10/2017 09:25, Johannes Morgenroth wrote:
>     The only way to "reserve" a port in a Unix/Linux system is to bind a
>     socket to that port.
> 
> As you may not know port ranges to use are easy to identify in Linux. 
> Usually, the ports above 32768 are used for dynamic allocations. The 
> port below are not used for other clients and are usually free. You can 
> check that using proc (even without super-user).
> 
> |$ cat /proc/sys/net/ipv4/ip_local_port_range 32768 60999 |

I'm quite aware about the ip_local_port_range parameter, which 
determines the port range used by the OS for dynamic port assignment. 
BTW its scope can also be observed (and modified with su privileges) 
with /sbin/sysctl net.ipv4.ip_local_port_range.

The problem is that although the OS dynamically assigns port numbers in 
that range, it does not prevent any Un*x process from also attempting to 
open a port in that range. Conversely, port numbers out of that range 
cannot be assumed to be free, since any process can open a socket on any 
port number at any time (except in the 0..1023 range): the only 
condition is that this port is not open yet.

This would not be a problem if the host running LEPTON (together with 
OppNet nodes) was entirely dedicated to that job. But since LEPTON and 
OppNet nodes are meant to run on a standard workstation (or a cluster 
node) they must share this host with dozens --if not hundreds-- of other 
Un*x processes. So we cannot assume that a given range of ports is 
available, and the reasonable approach is to assume that each OppNet 
node can request port numbers from the OS (using dynamic port assignment).

Now, as explained before, introducing dynamic port assignment in IBRDTN 
would be easy if a distinct socket was not created for each non-local 
address of an interface. Using interface "any" should be an option, but 
it seems this option is broken right now (see below).

> Getting determinism is an ambitious goal with real network stack and 
> emulated movement. Since the OppNets (at least IBR-DTN) is not really 
> deterministic due to multithreading, the determinism of your results 
> will be limited. 

Sure, but it will however be far better than what is obtained if the OS 
attempts to create (or terminate) all OppNet nodes simultaneously. As 
explained before the jitter observed on start and stop times often 
amounts to several seconds with "just-in-time" node creation or 
deletion, while it usually remains below a few tens of ms with our 
approach. A few tens of ms on the one hand, several seconds on the other 
hand: that's a significant difference.

> However, getting all nodes in sync at start is an good 
> point. But how do you find out the time needed to start your (thousands) 
> nodes? 10 seconds sounds like a empirical value.

10 seconds is indeed an empirical value, but it can be adjusted as 
desired (this is a parameter in LEPTON). Besides OppNet nodes are not 
asked to start 10 seconds after being created: they are asked to start 
at the exact time set in the command-line parameters (this is not a 
delay, it's a date, with ms precision). So OppNet nodes actually start 
running at the set time (or within a few ms of that time). The 10 --or 
whatever-- second advance delay simply provides some leeway for the 
system to create OppNet instances long before they are actually needed. 
It's even more important when running experiments with a cluster of 
workstations, since each OppNet must be started on a remote node (via 
ssh typically), which also takes time.

> I think there is a 
> better way to do it. In Hydra <https://github.com/morgenroth/hydrasim>, 
> the nodes are started without being connected to each other. As soon as 
> all nodes are ready, the discovery channel is opened. In you case, you 
> could easily control that channel using the discovery relay. The same 
> way, you could disconnect the nodes from the network. First detach and 
> then send a kill signal.

The problem is that LEPTON is meant to run a large variety of scenarios, 
including scenarios involving mobile networks with churn (i.e., all 
OppNet nodes are not necessarily created at time 0: nodes may actually 
be created and deleted all along an experiment). So opening the 
discovery channel at once for all nodes, and thus synchronizing all 
nodes, is not an option.
Besides LEPTON is also meant to run application-level scenarios on top 
of mobility or contact scenarios. Since all types of OppNet nodes do not 
operate the same way, we cannot assume, for example, that an application 
only starts sending messages (or beacons) once a transmission channel is 
available. Each level (i.e., beaconing, gossiping, application...) 
follows its own timeline, and respecting each timeline is important for 
the sake of repeatability.


>     Quite honestly I wonder why you don't simply create one socket
>     (listening to the desired port), without binding it to any interface
>     or address. Admittedly an IBRDTN may then receive messages (i.e.,
>     beacons or bundles) from the wrong interface, but would that be such
>     a problem?
> 
> This allows IBR-DTN to bind explicitly to specific interfaces with 
> different convergence-layers. If you want to bind to any interface (e.g. 
> *:4556), you can simply set "any" as interface.

As explained above, setting "any" as an interface would be great, but 
unfortunately it does not work.

I've tried several times inserting a line such as:

net_lan0_interface = any

... in a node's configuration file, rather than specifying the id of a 
specific interface. What I observe in that case is that the DTN daemon 
does not complain, but there is no communication whatsoever between that 
node and neighbor nodes (which run on other hosts and share the same 
Ethernet trunk).
With a packet sniffer such as Wireshark I observe that no beacon is sent 
by the node configured with interface 'any'. Besides with 'netstat -g 
-u' I can see that no group membership is registered for the multicast 
address that should be used for beaconing.

I've observed the same problems when using an argument such as '-i any' 
in dtnd's command line.

So, it seems the "any" feature is broken. It's unfortunate, because 
introducing support for dynamic port assignment would certainly be a lot 
simpler (with no need for the dummy socket "trick") with interface "any" 
than with a specific interface.

morgenroth commented:
I have fixed the bind to <any> interface issue and changed the default behavior of the daemon. Now, it will come up with a TCP-CL bound to <any> even if there is no configuration or "-i" parameter set.

```
$ dtnd
Sat Oct 14 15:14:58 2017 INFO NativeDaemon: IBR-DTN daemon 1.0.1 (build a171900)
Sat Oct 14 15:14:58 2017 INFO Configuration: Using default settings. Call with --help for options.
Sat Oct 14 15:14:58 2017 INFO NativeDaemon: Parallel event processing enabled using 4 processes.
Sat Oct 14 15:14:58 2017 INFO BundleCore: Local node name: dtn://link
Sat Oct 14 15:14:58 2017 INFO BundleCore: Forwarding of bundles enabled.
Sat Oct 14 15:14:58 2017 INFO BundleCore: Non-singleton bundles are accepted.
Sat Oct 14 15:14:58 2017 INFO NativeDaemon: using bundle storage in memory-only mode
Sat Oct 14 15:14:58 2017 INFO NativeDaemon: API initialized using tcp socket: <loopback>:4550
Sat Oct 14 15:14:58 2017 INFO NativeDaemon: TCP ConvergenceLayer added on <any>:4556
Sat Oct 14 15:14:58 2017 INFO IPNDAgent: listen to [ff02::142]:4551
Sat Oct 14 15:14:58 2017 INFO IPNDAgent: listen to [224.0.0.142]:4551
Sat Oct 14 15:14:58 2017 INFO IPNDAgent: advertise on interface <any>
Sat Oct 14 15:14:58 2017 INFO NativeDaemon: Using default routing extensions
```

```
$ netstat -gun
...
enp0s25         1      224.0.0.142
...
enp0s25         1      ff02::142
```
Maybe you can now figure out a more elegant way to apply your needs without the dummy socket "trick".

fguidec commented:
> I have fixed the bind to interface issue and changed the default 
> behavior of the daemon. Now, it will come up with a TCP-CL bound to even 
> if there is no configuration or "-i" parameter set.

Thank you for these modifications. I've tried running an IBRDTN node 
with the new code, and checked that this code indeed behaves as expected 
when using interface 'any'.

I should now be able to insert code supporting dynamic port assignment 
in TCPConvergenceLayer (around lines 122-125), without resorting to the 
"temporary socket trick".  ;-)

I'm still a bit confused by the behavior of the IPNDAgent when using 
interface 'any', though.

My concern is that several UDP sockets are still created in 
IPNDAgent::componentUp(), when I expected to see IPNDAgent running with 
a single socket (whose local port number could be easily assigned 
dynamically). More specifically:

- line 365: for each destination/discovery address, a socket is created 
and bound to 'any_addr' (that is, the wildcard address combined with the 
desired discovery port). This socket will therefore receive any beacon 
received on the specified port, whatever the destination address.

- line 403: for each interface, function join(iface) is invoked, which 
yields the creation of additional sockets (line 98) for each address 
associated with the interface.
Since function vinterface::getAddresses() returns two wildcard addresses 
for interface 'any', the outcome is that two additional sockets are 
created, the first one bound to (wildcard4, port), and the second one 
bound to (wildcard6, port). Note that if the configuration file 
specifies a single discovery address, and depending on whether this 
address is v4 or v6, one of these two sockets will never receive any 
beacon. Besides the remaining socket is redundant with the first one, 
created in line 365.

So, when setting interface 'any' in the configuration file (with a 
single discovery address), I've observed that three UDP sockets are 
indeed created.

$ sudo netstat -apu | grep dtnd
udp        0      0 0.0.0.0:4551    0.0.0.0:*        10217/dtnd
udp        0      0 0.0.0.0:4551    0.0.0.0:*        10217/dtnd
udp6       0      0 [::]:4551       [::]:*           10217/dtnd


IMHO two ot these three sockets are redundant, as the first one (created 
in line 365) should do the job...

Now, this multiplicity of sockets listening to the same port makes it 
difficult to support dynamic port assignment. Indeed, when 
discovery_port=0 we need to make sure that the port number assigned by 
the OS to the first socket created is then used to create all the 
remaining sockets (hence the "temporary socket trick", or something 
similar).

I would frankly prefer to get rid of this trick, but I do not see how to 
do it while the IPNDAgent creates so many sockets to listen to a single 
port.



-- 
Frédéric GUIDEC -----------------------------------------
Laboratoire IRISA
Université Bretagne Sud
Campus de Tohannic
BP 573 - F-56017 Vannes, France
mailto:Frederic.Guidec%AT%univ-ubs.fr
http://people.irisa.fr/Frederic.Guidec
tel : +33 (0)2 97 01 72 16
fax : +33 (0)2 97 01 72 79
---------------------------------------------------------

morgenroth commented:
If I start ibrdtn without any parameter port 4551 is only bound twice. For IPv6 and for IPv4. This is necessary, because different discovery messages are generated for the different protocols. Moreover, you can not bind multicast IPv4 on a IPv6 socket and vice-verca.

```
$ sudo netstat -apu | grep dtnd
udp        0      0 *:4551                  *:*                                 17134/dtnd
udp6       0      0 [::]:4551               [::]:*                              17134/dtnd
```

So, I think it is your configuration which causes the triple bind. Can you play with it to identify the responsible option.

fguidec commented:
> $ sudo netstat -apu | grep dtnd
> udp        0      0 *:4551                  *:*                                 17134/dtnd
> udp6       0      0 [::]:4551               [::]:*                              17134/dtnd
> 
> So, I think it is your configuration which causes the triple bind. Can you play with it to identify the responsible option.

Interesting... What I observe is quite different. FYI my workstation is running Linux Stretch, with a single Ethernet interface.

When running the latest ibrdtn/ibrdtn code, with no command-line option and no configuration file, I observe this:

```
$ sudo netstat -aup | grep dtnd
udp        0      0 0.0.0.0:4551            0.0.0.0:*                           5425/dtnd           
udp        0      0 0.0.0.0:4551            0.0.0.0:*                           5425/dtnd           
udp6       0      0 [::]:4551               [::]:*                              5425/dtnd           
udp6       0      0 [::]:4551               [::]:*                              5425/dtnd 

```
This is actually consistent with the current code:

1) since no discovery address is specified, two addresses (224.0.0.142 and ff02::142) are used by default, as planned in NativeDaemon (lines 1517-1529).
2) for each of these addresses, a socket is created in IPNDAgent, and this socket is bound to a wildcard address, with the same protocol family as the discovery address considered (line 365). So a socket bound to 0.0.0.0 is created because of group 224.0.0.2, and another socket bound to :: is created because of group ff02::142.
3) each interface is then "joined" in line 403, which yields the creation of additional sockets for each non-local address of each interface (line 98). When interface 'any' is considered by default, two additional sockets are thus created, one bound to 0.0.0.0, and one bound to ::.

So what I observe is consistent with the current code: two sockets are first created in step 2), and then two additional sockets are created in step 3).
The reason why I got only three sockets yesterday is that my configuration file defined a single IPv4 discovery address. So one socket was created in step 2), and two additional sockets in step 3).

The code that corresponds to step 2) (lines 349-384 in IPNDAgent) is conditional code that should be compiled only when \_\_WIN32\_\_ is not defined. Maybe that explains why we are not observing the same thing: are you working on a Windows system, by any chance?

> If I start ibrdtn without any parameter port 4551 is only bound twice. For IPv6 and for IPv4. This is necessary, because different discovery messages are generated for the different protocols. 

Actually the beacons sent to either 224.0.0.142 and ff02::142 are quite similar, as far as I can tell  (checked with Wireshark). Of course the IP headers are different, but the UDP datagrams themselves have the same structure and content. More specifically the service block in all beacons is similar, since CLA-TCP-HN  tags are used by default in IBRDTN (for TCP-CL, that is).

> Moreover, you can not bind multicast IPv4 on a IPv6 socket and vice-verca.

Actually, a UDP socket bound to the IPv6 wildcard address can equally receive datagrams sent to IPv4 or IPv6 multicast groups, provided option IPV6_V6ONLY is disabled (which is the default in Linux).
Anyway, even if you find it cleaner to create and maintain one socket bound to 0.0.0.0, and another one bound to ::, these two sockets are sufficient to receive any beacon, whatever its destination address. Of course it is still necessary to join each multicast group considered in the configuration parameters, but this can be done on these two sockets. There is no need to create any additional socket. 

Now, I'm going to try inserting dynamic port assignment again, in the present code. I'm pretty confident it can now be done without using the "temporary socket trick" in TCP-CL, and I hope I can also avoid using this trick in IPNDAgent, in spite of the multiplicity of sockets. Wish me luck!

morgenroth commented:
> So what I observe is consistent with the current code: two sockets are first created in step 2), and then two additional sockets are created in step 3).

Although I did test this in Linux, there are only 2 sockets. However, I agree that the code in step 2 might be redundant. I have to test if that is working on all platforms.

>> Moreover, you can not bind multicast IPv4 on a IPv6 socket and vice-verca.

> Actually, a UDP socket bound to the IPv6 wildcard address can equally receive datagrams sent to IPv4 or IPv6 multicast groups, provided option IPV6_V6ONLY is disabled (which is the default in Linux).

Sorry, what I wanted to say is that you can not send IPv4 broadcast from IPv6 socket and vice versa. I thought it should work, but I tried and it didn't. However, it is still required to have control over the used source addresse (IPv4 or IPv6) to generate specialized messages.

For your proposal, keep in mind that it is not necessary to have the same port for all IPNDAgent sockets. If you apply random ports to all of them it should work fine.

fguidec commented:
I observe that a lot of things have changed in ibrdtn:master in the last few days.

Indeed some of these changes now make it simpler to support dynamic port assignment in [IPNDAgent](https://github.com/fguidec/ibrdtn/blob/35f5e8c38d0814ebba770347ab8842b48ab545a7/ibrdtn/daemon/src/net/IPNDAgent.cpp#L407) and [TCPConvergenceLayer](https://github.com/fguidec/ibrdtn/blob/35f5e8c38d0814ebba770347ab8842b48ab545a7/ibrdtn/daemon/src/net/TCPConvergenceLayer.cpp#L129). Most notably, the "dummy socket trick" is not required anymore!
